Fix `auto_build.py` inconsistencies

- When sequential build fails, it can still print `"All module builds completed successfully!"`. That happens because when `sys.exit(1)`is called on the `die()`function, it only terminates the thread that calls it, but the main thread keeps running.
  ```
  def die(msg: str):
      print(f"Error: {msg}", file=sys.stderr)
      sys.exit(1)
  ```
  Which means that the main thread eventually prints success, even after a failure.
  ```
  # Sequential group (runs in background)
  if sequential:
  
      def run_sequential():
          print(f"Starting sequential module builds:{' '.join(sequential)}")
          for f in sequential:
              build_module(f, quiet) <-- build fails, die() is called and sequential thread dies quietly
  
      import threading
  
      seq_thread = threading.Thread(target=run_sequential)
      seq_thread.start()
  
  # Parallel builds
  [...]
  
  if sequential:
      seq_thread.join()
  
  print("All module builds completed successfully!") <-- main thread still prints success
  ```
  Example building secp256k1:
  ```
  CC=clang CXX=clang++ CXXFLAGS="-DSECP256K1" ./auto_build.py
  Cleaning previous builds...
  make clean
  rm -rf *.o module.a bitcoinfuzz include/bitcoinfuzz/*.o helpers/*.o modules/secp256k1/module.a
  rm -rf modules/eclair/eclair.zip modules/eclair/lib modules/eclair/eclair_extracted
  No CLEAN_BUILD option specified. Skipping clean step.
  Compiling selected modules in parallel (jobs=0) with CXXFLAGS=-DSECP256K1...
  Starting sequential module builds:SECP256K1
  cd ../../external/secp256k1 && \
  ./autogen.sh && \
  ./configure \
          --enable-module-recovery \
          --enable-experimental \
          --enable-module-schnorrsig \
          --disable-shared \
          --with-pic \
          CFLAGS="-fsanitize=address,fuzzer-no-link" && \
  make
  
  /bin/sh: line 2: ./autogen.sh: No such file or directory
  make: *** [Makefile:15: ../../external/secp256k1/.libs/libsecp256k1.a] Error 127
  
  Error: Command failed: make                            <-- build unsuccessful
  All module builds completed successfully!              <-- success build message
  Compiling the main project in the root...
  make
  [...]
  ```
- Automatic script assumes that necessary submodules are already initialized, which may cause build to fail on a freshly cloned repo:
   ```
  CC=clang CXX=clang++ CXXFLAGS="-DSECP256K1" ./auto_build.py
  Cleaning previous builds...
  make clean
  rm -rf *.o module.a bitcoinfuzz include/bitcoinfuzz/*.o helpers/*.o modules/secp256k1/module.a
  rm -rf modules/eclair/eclair.zip modules/eclair/lib modules/eclair/eclair_extracted
  No CLEAN_BUILD option specified. Skipping clean step.
  Compiling selected modules in parallel (jobs=0) with CXXFLAGS=-DSECP256K1...
  Starting sequential module builds:SECP256K1
  cd ../../external/secp256k1 && \
  ./autogen.sh && \
  ./configure \
          --enable-module-recovery \
          --enable-experimental \
          --enable-module-schnorrsig \
          --disable-shared \
          --with-pic \
          CFLAGS="-fsanitize=address,fuzzer-no-link" && \
  make
  
  /bin/sh: line 2: ./autogen.sh: No such file or directory    <-- Missing autogen.sh due to unitialized submodule
  [...]
  ```

--- a/auto_build.py
+++ b/auto_build.py
@@ -133,14 +133,20 @@ def main():
     parallel = [f for f in flags if not should_build_sequentially(f)]
 
     seq_pid = None
+    seq_error = None
 
     # Sequential group (runs in background)
     if sequential:
 
         def run_sequential():
+            nonlocal seq_error
             print(f"Starting sequential module builds:{' '.join(sequential)}")
-            for f in sequential:
-                build_module(f, quiet)
+            try:
+                for f in sequential:
+                    build_module(f, quiet)
+            except BaseException as e:
+                # Captures SystemExit raised by die()'s sys.exit()
+                seq_error = e
 
         import threading
 
@@ -161,6 +167,9 @@ def main():
 
     if sequential:
         seq_thread.join()
+        if seq_error:
+            # Re-raise/abort in the main thread so the script actually fails.
+            die("Sequential module builds failed")
 
     print("All module builds completed successfully!")
 

--- a/auto_build.py
+++ b/auto_build.py
@@ -63,6 +63,22 @@ def should_build_sequentially(flag: str) -> bool:
 def get_flags(cxxflags: str):
     return re.findall(r"-D([A-Z0-9_]+)", cxxflags)
 
+# Maps module flags to required git submodule paths defined in .gitmodules.
+SUBMODULES_BY_FLAG = {
+    "CLIGHTNING": ["external/lightning"],
+    "SECP256K1": ["external/secp256k1"],
+    "ECLAIR": ["external/eclair"],
+}
+
+def ensure_submodules_for(flag: str, quiet: bool):
+    paths = SUBMODULES_BY_FLAG.get(flag, [])
+    if not paths:
+        return
+    if not quiet:
+        print(f"Ensuring submodules for {flag}: {', '.join(paths)}")
+    for path in paths:
+        run(f"git submodule update --init --recursive {path}", quiet=quiet)
+
 def full_clean():
     print("Performing full clean...")
     for d in Path("modules").glob("*/"):
@@ -79,6 +95,8 @@ def build_module(flag: str, quiet: bool):
 
     if not quiet:
         print(f"Building module: {flag}")
+
+    ensure_submodules_for(flag, quiet)
 
     if needs_rust_nightly(flag):
         execute_in_dir(

